File: bagtrip/lib/models/traveler_profile.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'traveler_profile.freezed.dart';
part 'traveler_profile.g.dart';

@freezed
abstract class TravelerProfile with _$TravelerProfile {
  const factory TravelerProfile({
    required String id,
    @JsonKey(name: 'travelTypes') @Default([]) List<String> travelTypes,
    @JsonKey(name: 'travelStyle') String? travelStyle,
    String? budget,
    String? companions,
    @JsonKey(name: 'travelFrequency') String? travelFrequency,
    @JsonKey(name: 'medicalConstraints') String? medicalConstraints,
    @JsonKey(name: 'isCompleted') @Default(false) bool isCompleted,
    @JsonKey(name: 'createdAt') DateTime? createdAt,
    @JsonKey(name: 'updatedAt') DateTime? updatedAt,
  }) = _TravelerProfile;

  factory TravelerProfile.fromJson(Map<String, dynamic> json) =>
      _$TravelerProfileFromJson(json);
}

@freezed
abstract class ProfileCompletion with _$ProfileCompletion {
  const factory ProfileCompletion({
    @JsonKey(name: 'isCompleted') @Default(false) bool isCompleted,
    @JsonKey(name: 'missingFields') @Default([]) List<String> missingFields,
  }) = _ProfileCompletion;

  factory ProfileCompletion.fromJson(Map<String, dynamic> json) =>
      _$ProfileCompletionFromJson(json);
}

File: api/src/models/traveler_profile.py
"""Modèle TravelerProfile SQLAlchemy."""

import uuid
from datetime import datetime
from uuid import UUID as _UUID

from sqlalchemy import Boolean, DateTime, ForeignKey, String
from sqlalchemy.dialects.postgresql import JSON, UUID
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.sql import func

from src.config.database import Base


class TravelerProfile(Base):
    """Profil voyageur avec préférences de personnalisation."""

    __tablename__ = "traveler_profiles"

    id: Mapped[_UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[_UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), unique=True, nullable=False, index=True
    )
    travel_types: Mapped[list | None] = mapped_column(JSON, nullable=True)
    travel_style: Mapped[str | None] = mapped_column(String, nullable=True)
    budget: Mapped[str | None] = mapped_column(String, nullable=True)
    companions: Mapped[str | None] = mapped_column(String, nullable=True)
    medical_constraints: Mapped[str | None] = mapped_column(String, nullable=True)
    travel_frequency: Mapped[str | None] = mapped_column(String, nullable=True)
    is_completed: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False
    )
